### PR TITLE
Docstring108

### PR DESCRIPTION
**Pull request by WChrisK**
There can be some cleanup of the code, and there's some TODO's, but its fully functional and the tests rigorously show that it does.

I still think it can be cleaner however, and I finished this because `funcparserlib` was taking a significant amount of time to understand and I didn't want to be emptyhanded.

### CODE REVIEW

**Comment by david-yz-liu** [comment]
Thanks, @WChrisK. I am going to close the other pull request.

**Comment by WChrisK** [comment]
Going to update funcparserlib stuff to be better.

File: docstring/csc108_docstring_fpl.py [hunk 1]
```diff
@@ -0,0 +1,70 @@
+from funcparserlib.parser import a, forward_decl, maybe, finished, many, skip, oneplus, some, _Tuple
+
+
+# Precondition: 's' is a string with at least one character
+def a_str(s):
+    p_out = a(s[0])
+    for i in range(1, len(s)):
+        p_out += a(s[i])
+    return (p_out >> to_str)
```
Comment (david-yz-liu) [inline docstring/csc108_docstring_fpl.py]: Don't use `>>`

File: docstring/csc108_docstring_fpl.py [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+
+def skip_str(s):
+    return skip(a_str(s))
+
+def to_str(s):
+    return ''.join(s)
+
+# Precondition: args length is >= 1
+def inject_whitespace(*args):
+	if len(args) == 1:
```
Comment (david-yz-liu) [inline docstring/csc108_docstring_fpl.py]: It appears that your editor again is using tabs.

File: docstring/csc108_docstring_fpl.py [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+		out += ws + args[i]
+	return out + ws + args[-1]
+
+def parse_docstring(docstr):
+	return docstring.parse(docstr)
+
+element = forward_decl()
+elements = forward_decl()
+ws = skip(many(a(' ') | a('\t')))
+anystr = oneplus(some(lambda c: c.isalpha())) >> to_str
```
Comment (david-yz-liu) [inline docstring/csc108_docstring_fpl.py]: You should be able to name the function as `str.isalpha` or something

File: docstring/csc108_docstring.py [hunk 4]
```diff
@@ -0,0 +1,288 @@
...
+                tuple of (int, int)
+    This is needed since PEP-8 being enforced in CSC108 means the possibility of
+    the docstring being greater than 80 characters, and it must be supported if
+    its too long.
+    This function will properly extract the string and compress it into a single
+    string with all newlines/improper spacing removed.
+    Note: This does not check for correctness of the docstring header contract,
+    its job is only to extract the docstring, and let the parser error out if
+    the type contract is malformed.
+    """
```
Comment (nigef) [inline docstring/csc108_docstring.py]: nice writeup 👌

**Comment by david-yz-liu** [comment]
@WChrisK I merged this in, but there's an outstanding comment in this pull request you need to fix.

Also, you should look up how to use `assertRaises` in unittest to improve your tests.

**Pull request merged**